File: src/823. Binary Trees With Factors.js
/**
 * @param {number[]} arr
 * @return {number}
 */

export function numFactoredBinaryTrees(arr) {
  arr.sort((a, b) => a - b)
  const map = new Map(arr.map((product) => [product, 1])),
    mod = 10 ** 9 + 7
  for (let product of arr) {
    for (let value of arr) {
      const quotient = product / value
      if (quotient < 1) break
      if (!map.has(quotient)) continue
      map.set(product, map.get(product) + map.get(value) * map.get(quotient))
    }
  }
  return arr.reduce((sum, product) => sum + map.get(product), 0) % mod
}

File: src/823. Binary Trees With Factors.py
class Solution:
    def numFactoredBinaryTrees(self, arr: List[int]) -> int:
        map = {}
        mod = 10 ** 9 + 7
        for product in sorted(arr):
            map[product] = 1 + sum(
                map[value] * map.get(product / value, 0)
                for value in arr if value < product
            )
        return sum(map.values()) % mod
